src/Google.Datastore.V1Beta3/DatastoreDb.cs

```diff
@@ -119,34 +125,49 @@ public class DatastoreDb
         /// using default settings.</param>
         /// <returns>A <see cref="DatastoreDb"/> operating on the specified partition.</returns>
         public static DatastoreDb Create(string projectId, string namespaceId = "", DatastoreClient client = null) =>
-            new DatastoreDb(projectId, namespaceId, client ?? DatastoreClient.Create());
-
-        private DatastoreDb(string projectId, string namespaceId, DatastoreClient client)
-        {
-            _projectId = GaxPreconditions.CheckNotNull(projectId, nameof(projectId));
-            _partitionId = new PartitionId(projectId, GaxPreconditions.CheckNotNull(namespaceId, nameof(namespaceId)));
-            Client = GaxPreconditions.CheckNotNull(client, nameof(client));
-        }
+            new DatastoreDbImpl(projectId, namespaceId, client ?? DatastoreClient.Create());
 
         /// <summary>
         /// Creates a key factory for root entities in this objects's partition.
         /// </summary>
         /// <param name="kind">The kind of entity key to create. Must not be null.</param>
         /// <returns>A key factory with the specified kind and this object's partition.</returns>
-        public KeyFactory CreateKeyFactory(string kind) => new KeyFactory(_partitionId, kind);
+        public virtual KeyFactory CreateKeyFactory(string kind)
```

Comment:
On Tue, Jun 7, 2016 at 10:45 PM, Jon Skeet <notifications@github.com> wrote:

> In src/Google.Datastore.V1Beta3/DatastoreDb.cs
> <https://github.com/GoogleCloudPlatform/gcloud-dotnet/pull/193#discussion_r66195449>
> :
>
> >
> >          /// <summary>
> >          /// Creates a key factory for root entities in this objects's partition.
> >          /// </summary>
> >          /// <param name="kind">The kind of entity key to create. Must not be null.</param>
> >          /// <returns>A key factory with the specified kind and this object's partition.</returns>
> > -        public KeyFactory CreateKeyFactory(string kind) => new KeyFactory(_partitionId, kind);
> > +        public virtual KeyFactory CreateKeyFactory(string kind)
>
> Well fundamentally, I can't see how it could work with any API - if we
> assume that the project ID at least is required,
>
You spoke earlier of testing that assumption.  Have you tested it?


> and that in many cases a namespace would be used too. I'm still surprised
> that it worked without supplying those in the REST API... if you wanted to
> access entities in two different partitions, for example, you have to
> differentiate between them somewhere.
>
> Now just because there's a CreateKeyFactory method on DatastoreDb doesn't
> mean you can't create a KeyFactory other ways - if you're happy using a
> single project ID and namespace, you could create a KeyFactory and store
> it in a static variable, then use that in your extension method.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/GoogleCloudPlatform/gcloud-dotnet/pull/193/files/c6fb88e98632cb71f3d1d59df4e4a1a2ae4fb9e6#r66195449>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ALO54rhmj5CFNzVqTSJTP5Hm-smZB690ks5qJlcUgaJpZM4IvwZ0>
> .
>